logger/logger_writer.go

```diff
@@ -311,14 +311,22 @@ func Close(ctx context.Context) error {
 // Pending messages may never run out if another goroutine is constantly
 // writing.
 func DrainContext(ctx context.Context) error {
-	for ctx.Err() == nil && len(messages) > 0 {
+	for ctx.Err() == nil && (len(messages) > 0 || len(freeMessages) < cap(freeMessages)) {
 		innerCtx, cancel := context.WithTimeout(ctx, 10*time.Millisecond)
-		<-innerCtx.Done() // Wait for 10ms and check len(messages) again
+		<-innerCtx.Done() // Wait for 10ms and check the queues again
```

Comment:
I'm not seeing how this is different than just calling `time.Sleep(10*time.Millisecond)`.

I guess a good reason to have `innerCtx` would be to cancel the timeout when `ctx` is cancelled.  But that's not implemented here.  And I guess 10ms is short enough that we don't care.

Reply:
> I'm not seeing how this is different than just calling time.Sleep(10*time.Millisecond).

The outer context `ctx` might timeout (or be cancelled) sooner than 10ms.

> I guess a good reason to have innerCtx would be to cancel the timeout when ctx is cancelled. But that's not implemented here.

It is.  Context cancellation is transitive.  If the parent context `ctx` times out or is cancelled, `innerCtx` will automatically be cancelled too.

Reply:
oh shoot, I missed that detail.  Thanks for correcting me.